synIndex.R

```diff
@@ -25,44 +34,119 @@ source('params.R')
+## This would depend on the how the data is organized in the main S3 INGRESS bucket
+FOLDERS_TO_SYNC_SYNAPSE <- c('adults',
+                             'pregnant',
+                             'pediatric') 
+
+## From the local AWS location remove all folders that are not the ones selected above
+dirs_to_delete <- dplyr::setdiff(localDirs, c(FOLDERS_TO_SYNC_SYNAPSE, "")) # "" is the local directory, a result of list.dirs(.. full.names=FALSE)
+
+## delete the unwanted directories before creating a upload manifest
+for(dir_ in dirs_to_delete){
+  unlink(paste0(AWS_DOWNLOAD_LOCATION,"/",dir_), recursive = TRUE)
+}
+
+# localFileList <- list.files(path = AWS_DOWNLOAD_LOCATION,
+#                             all.files = TRUE, # get hidden files too
+#                             recursive = TRUE, # get files inside sub-folders (if any)
+#                             full.names = FALSE) # to get the directory path prepended to the file name
+
+#############
+# Get a manifest of files to upload
+#############
+## The S3 bucket is synced to AWS_DOWNLOAD_LOCATION locally. 
+## We will use synapse cmd line client manifest function to replicate the folder structure 
+## but NOT upload the files. We will create a datafilehandleid in place later instead of uploading file
+
+old_path <- Sys.getenv("PATH")
+Sys.setenv(PATH = paste(old_path, "/home/ubuntu/R/x86_64-pc-linux-gnu-library/3.6/PythonEmbedInR/bin", sep = ":"))
+## When we install synapser, it installs synapseclient and along with it the synapse cmd line client
+## We need to add the location of synapseclient to the system path so that it can recognize synapse cmd 
+
+SYSTEM_COMMAND <- glue::glue('synapse -u "{SYNAPSE_USERNAME}" -p "{SYNAPSE_PASSWORD}" manifest --parent-id {SYNAPSE_PARENT_ID} --manifest current_manifest.tsv {AWS_DOWNLOAD_LOCATION}')
+
+## Generate manifest file
+system(SYSTEM_COMMAND)
+
+###########
+## Get a list of all files to upload and their synapse locations(parentId) 
+###########
+STR_LEN_AWS_DOWNLOAD_LOCATION = stringr::str_length(AWS_DOWNLOAD_LOCATION)
+
+## All files present locally from manifest
+synapse_manifest <- read.csv('current_manifest.tsv', sep = '\t', stringsAsFactors = F) %>% 
+  dplyr::filter(path != paste0(AWS_DOWNLOAD_LOCATION,'/owner.txt')) %>%  # need not create a dataFileHandleId for owner.txt
+  dplyr::rowwise() %>% 
+  dplyr::mutate(file_key = stringr::str_sub(string = path, start = STR_LEN_AWS_DOWNLOAD_LOCATION+2)) %>% # location of file from home folder of S3 bucket 
+  dplyr::ungroup()
+
+## All currently indexed files in Synapse
+synapse_fileview <- synapser::synTableQuery(paste0('SELECT * FROM ', SYNAPSE_FILEVIEW_ID))$asDataFrame()
+
+## find those files that are not in the fileview - files that need to be indexed
+synapse_manifest_to_upload <- synapse_manifest %>% 
+  dplyr::anti_join(synapse_fileview %>% 
+                     dplyr::select(parent = parentId,
+                                   file_key = dataFileKey))
 
```

Comment:
What about files that have changed.  I am not sure this is possible but if there are changes how should those be handled.  Should also be ropbust to unexpected changes.  Maybe use md5 as an additional check.

In general, expect the unexcpted and fail fail fail loudly when the unexpected happens.

Reply:
Added additional md5 check.